pkg/endpoint/events.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+import (
+	"sync"
+
+	"github.com/cilium/cilium/pkg/eventqueue"
+)
+
+// EndpointRegenerationEvent contains all fields necessary to regenerate an endpoint.
+type EndpointRegenerationEvent struct {
+	owner        Owner
+	regenContext *regenerationContext
+	ep           *Endpoint
+}
+
+// Handle handles the regeneration event for the endpoint.
+func (ev *EndpointRegenerationEvent) Handle(res chan interface{}) {
+	e := ev.ep
+	owner := ev.owner
+	regenContext := ev.regenContext
+
+	err := e.RLockAlive()
+	if err != nil {
+		e.LogDisconnectedMutexAction(err, "before regeneration")
+		res <- &EndpointRegenerationResult{
+			err: err,
+		}
+
+		return
+	}
+	e.RUnlock()
+
+	// We should only queue the request after we use all the endpoint's
+	// lock/unlock. Otherwise this can get a deadlock if the endpoint is
+	// being deleted at the same time. More info PR-1777.
+	doneFunc := owner.QueueEndpointBuild(uint64(e.ID))
+	if doneFunc != nil {
+		e.getLogger().Debug("Dequeued endpoint from build queue")
+
+		regenContext.DoneFunc = doneFunc
+
+		err = ev.ep.regenerate(ev.owner, ev.regenContext)
+
+		doneFunc()
+		e.notifyEndpointRegeneration(owner, err)
+	} else {
+		e.getLogger().Debug("My request was cancelled because I'm already in line")
+	}
+
+	res <- &EndpointRegenerationResult{
+		err: err,
+	}
+	return
+}
+
+// EndpointRegenerationResult contains the results of an endpoint regeneration.
+type EndpointRegenerationResult struct {
+	err error
+}
+
+// EndpointRevisionBumpEvent contains all fields necessary to bump the policy
+// revision of a given endpoint.
+type EndpointRevisionBumpEvent struct {
+	Rev uint64
+	ep  *Endpoint
+}
+
+// Handle handles the revision bump event for the Endpoint.
+func (ev *EndpointRevisionBumpEvent) Handle(res chan interface{}) {
+	// TODO: if the endpoint is not in a 'ready' state that means that
+	// we cannot set the policy revision, as something else has
+	// changed endpoint state which necessitates regeneration,
+	// *or* the endpoint is in a not-ready state (i.e., a prior
+	// regeneration failed, so there is no way that we can
+	// realize the policy revision yet. Should this be signaled
+	// to the routine waiting for the result of this event?
+	ev.ep.SetPolicyRevision(ev.Rev)
+	res <- struct{}{}
```

Comment:
Should this be passing back an `EndpointRegenerationResult`?

Reply:
No, this is an `EndpointRevisionBumpEvent`.